wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java

```diff
@@ -96,37 +99,189 @@ public LimitlessRouterServiceImpl(
+        continue;
+      }
+
+      try {
+        context.setConnection(pluginService.connect(selectedHostSpec, context.getProps()));
+        if (context.getConnection() != null) {
+          return;
+        }
+      } catch (final SQLException e) {
+        selectedHostSpec.setAvailability(HostAvailability.NOT_AVAILABLE);
+        LOGGER.finest(Messages.get(
+            "LimitlessRouterServiceImpl.failedToConnectToHost",
+            new Object[] {selectedHostSpec.getHost()}));
+      }
+    }
+    throw new SQLException(Messages.get("LimitlessRouterServiceImpl.maxRetriesExceeded"));
+  }
+
+  protected void synchronouslyGetLimitlessRoutersWithRetry(final LimitlessConnectionContext context)
+      throws SQLException {
+    LOGGER.finest(Messages.get("LimitlessRouterServiceImpl.synchronouslyGetLimitlessRouters"));
+    int retryCount = -1; // start at -1 since the first try is not a retry.
+    int maxRetries = LimitlessConnectionPlugin.GET_ROUTER_MAX_RETRIES.getInteger(context.getProps());
+    int retryIntervalMs = LimitlessConnectionPlugin.GET_ROUTER_RETRY_INTERVAL_MILLIS.getInteger(context.getProps());
+    do {
+      try {
+        synchronouslyGetLimitlessRouters(context);
+        if (!Utils.isNullOrEmpty(context.getLimitlessRouters())) {
+          return;
+        }
+        Thread.sleep(retryIntervalMs);
+      } catch (final InterruptedException e) {
+        Thread.currentThread().interrupt();
+      } finally {
+        retryCount++;
+      }
+    } while (retryCount < maxRetries);
+    throw new SQLException(Messages.get("LimitlessRouterServiceImpl.noRoutersAvailable"));
+  }
+
+  protected void synchronouslyGetLimitlessRouters(final LimitlessConnectionContext context)
+      throws SQLException {
     final long cacheExpirationNano = TimeUnit.MILLISECONDS.toNanos(
-        MONITOR_DISPOSAL_TIME_MS.getLong(props));
+        MONITOR_DISPOSAL_TIME_MS.getLong(context.getProps()));
 
     final ReentrantLock lock = forceGetLimitlessRoutersLockMap.computeIfAbsent(
         this.pluginService.getHostListProvider().getClusterId(),
-        (key) -> new ReentrantLock()
+        key -> new ReentrantLock()
     );
     lock.lock();
     try {
       final List<HostSpec> limitlessRouters =
           limitlessRouterCache.get(this.pluginService.getHostListProvider().getClusterId(), cacheExpirationNano);
       if (!Utils.isNullOrEmpty(limitlessRouters)) {
-        return limitlessRouters;
+        context.setLimitlessRouters(limitlessRouters);
+        return;
       }
 
+      if (context.getConnection() == null || context.getConnection().isClosed()) {
+        context.setConnection(context.getConnectFunc().call());
+      }
       final List<HostSpec> newLimitlessRouters =
-          this.queryHelper.queryForLimitlessRouters(connection, hostPort);
+          this.queryHelper.queryForLimitlessRouters(context.getConnection(), context.getHostSpec().getPort());
 
+      context.setLimitlessRouters(newLimitlessRouters);
```

Comment:
if (!Utils.isNullOrEmpty(newLimitlessRouters)) {
...
}

Reply:
Added a line to throw an exception in case the it is either null or empty. 
Reason for this is so that `#synchronouslyGetLimitlessRoutersWithRetry` can catch it and retry. 
Otherwise if `context.limitlessRouters` is not null or empty for some reason, it will not retry.